File: event-aggregator/src/test/scala/com/boclips/event/aggregator/testsupport/testfactories/ContractFactory.scala
package com.boclips.event.aggregator.testsupport.testfactories

import java.util.Currency

import com.boclips.event.aggregator.domain.model.ContractLegalRestriction
import com.boclips.event.aggregator.domain.model.contentpartners._
import com.boclips.event.aggregator.presentation.formatters.schema.base.ExampleInstance
import com.boclips.event.aggregator.presentation.model.ContractTableRow

object ContractFactory {
  def createFullContract(
                          id: String = "contract-id",
                          channelName: String = "channel name",
                          contractDocumentLink: String = "http://google.com",
                          contractIsRolling: Boolean = true,
                          contractDates: ContractDates = ExampleInstance.create[ContractDates],
                          daysBeforeTerminationWarning: Int = 30,
                          yearsForMaximumLicense: Int = 90,
                          daysForSellOffPeriod: Int = 100,
                          royaltySplit: ContractRoyaltySplit = ExampleInstance.create[ContractRoyaltySplit],
                          minimumPriceDescription: String = "minimum",
                          remittanceCurrency: Currency = Currency.getInstance("GBP"),
                          restrictions: ContractRestrictions = ExampleInstance.create[ContractRestrictions],
                          costs: ContractCosts = ExampleInstance.create[ContractCosts]
                        ): Contract =
    Contract(
      id = ContractId(id),
      name = channelName,
      contractDocumentLink = Some(contractDocumentLink),
      contractIsRolling = Some(contractIsRolling),
      contractDates = Some(contractDates),
      daysBeforeTerminationWarning = Some(daysBeforeTerminationWarning),
      yearsForMaximumLicense = Some(yearsForMaximumLicense),
      daysForSellOffPeriod = Some(daysForSellOffPeriod),
      royaltySplit = Some(royaltySplit),
      minimumPriceDescription = Some(minimumPriceDescription),
      remittanceCurrency = Some(remittanceCurrency),
      restrictions = Some(restrictions),
      costs = costs
    )

  def createEmptyContract(
                           id: String = "contract-id",
                           channelName: String = "channel name"
                         ): Contract =
    Contract(
      id = ContractId(id),
      name = channelName,
      contractDocumentLink = None,
      contractIsRolling = None,
      contractDates = None,
      daysBeforeTerminationWarning = None,
      yearsForMaximumLicense = None,
      daysForSellOffPeriod = None,
      royaltySplit = None,
      minimumPriceDescription = None,
      remittanceCurrency = None,
      restrictions = None,
      costs = ContractCosts(
        minimumGuarantee = List(),
        upfrontLicense = None,
        technicalFee = None,
        recoupable = None
      )
    )

  def createContractRestrictions(
                                  clientFacing: Option[List[String]] = None,
                                  territory: Option[String] = None,
                                  licensing: Option[String] = None,
                                  editing: Option[String] = None,
                                  marketing: Option[String] = None,
                                  companies: Option[String] = None,
                                  payout: Option[String] = None,
                                  other: Option[String] = None,
                                ): ContractRestrictions =
    ContractRestrictions(
      clientFacing = clientFacing,
      territory = territory,
      licensing = licensing,
      editing = editing,
      marketing = marketing,
      companies = companies,
      payout = payout,
      other = other
    )

  def createContractRestriction(
                                 id: String = "id-998",
                                 text: String = "rules allowed"
                               ): ContractLegalRestriction =
    ContractLegalRestriction(
      id = id,
      text = text
    )

  def createFullTableRowContract(
                                  contract: Contract = createFullContract(),
                                  clientFacingRestrictions: List[ContractLegalRestriction] = List(createContractRestriction())

  ) : ContractTableRow =
    ContractTableRow(
      contract = contract,
      clientFacingRestrictions = clientFacingRestrictions
    )
}

File: event-service/src/test/kotlin/com/boclips/event/service/testsupport/ContractFactory.kt
package com.boclips.event.service.testsupport

import com.boclips.eventbus.domain.contract.Contract
import com.boclips.eventbus.domain.contract.ContractCosts
import com.boclips.eventbus.domain.contract.ContractDates
import com.boclips.eventbus.domain.contract.ContractId
import com.boclips.eventbus.domain.contract.ContractRestrictions
import com.boclips.eventbus.domain.contract.ContractRoyaltySplit
import java.math.BigDecimal
import java.time.LocalDate
import java.time.Month
import java.util.Currency

object ContractFactory {
    fun createContract(
        id: String = "contract-id",
        name: String = "contract-name",
        contractDocument: String = "http://contractdocument.com",
        contractIsRolling: Boolean = true,
        contractDates: ContractDates = createContractDates(),
        daysBeforeTerminationWarning: Int = 30,
        yearsForMaximumLicense: Int = 60,
        daysForSellOffPeriod: Int = 90,
        royaltySplit: ContractRoyaltySplit = createContractRoyaltySplit(),
        minimumPriceDescription: String = "minimum price description",
        remittanceCurrency: Currency = Currency.getInstance("GBP"),
        restrictions: ContractRestrictions = createContractRestrictions(),
        costs: ContractCosts = createContractCosts()
    ): Contract =
        Contract.builder()
            .contractId(ContractId(id))
            .name(name)
            .contractDocument(contractDocument)
            .contractIsRolling(contractIsRolling)
            .contractDates(contractDates)
            .daysBeforeTerminationWarning(daysBeforeTerminationWarning)
            .yearsForMaximumLicense(yearsForMaximumLicense)
            .daysForSellOffPeriod(daysForSellOffPeriod)
            .royaltySplit(royaltySplit)
            .minimumPriceDescription(minimumPriceDescription)
            .remittanceCurrency(remittanceCurrency)
            .restrictions(restrictions)
            .costs(costs)
            .build()

    fun createContractDates(
        start: LocalDate = LocalDate.of(2020, Month.DECEMBER, 31),
        end: LocalDate = LocalDate.of(2022, Month.JANUARY, 1)
    ): ContractDates =
        ContractDates.builder()
            .start(start)
            .end(end)
            .build()

    fun createContractRoyaltySplit(
        download: Float = 50F,
        streaming: Float = 60.3F
    ): ContractRoyaltySplit =
        ContractRoyaltySplit.builder()
            .download(download)
            .streaming(streaming)
            .build()

    fun createContractRestrictions(
        clientFacing: List<String> = listOf("client-facing"),
        territory: String = "territory",
        licensing: String = "licensing",
        editing: String = "editing",
        marketing: String = "marketing",
        companies: String = "companies",
        payout: String = "payout",
        other: String = "other"
    ): ContractRestrictions =
        ContractRestrictions.builder()
            .clientFacing(clientFacing)
            .territory(territory)
            .licensing(licensing)
            .editing(editing)
            .marketing(marketing)
            .companies(companies)
            .payout(payout)
            .other(other)
            .build()

    fun createContractCosts(
        minimumGuarantee: List<BigDecimal> = listOf(BigDecimal.TEN),
        upfrontLicense: BigDecimal = BigDecimal.ONE,
        technicalFee: BigDecimal = BigDecimal.ZERO,
        recoupable: Boolean = true
    ): ContractCosts =
        ContractCosts.builder()
            .minimumGuarantee(minimumGuarantee)
            .upfrontLicense(upfrontLicense)
            .technicalFee(technicalFee)
            .recoupable(recoupable)
            .build()
}